content/blog/2021-07-14-continuous-benchmarking.md

```diff
@@ -0,0 +1,144 @@
+
+[^2]: Technically you can still benchmarks on the CI by measuring a different metric. See [benchmarking](https://pythonspeed.com/articles/consistent-benchmarking-in-ci/) in [noisy](https://bheisler.github.io/post/benchmarking-in-the-cloud/) [environments](https://bheisler.github.io/post/criterion-rs-0-3-4/).
+
+Secondly, we need someplace to store that benchmark data so that we can view
+how it changed over time. This can be a SQL database, a metrics oriented
+database or just a simple git repository. We store our benchmark data in a JSON
+file and push it to a git repository.
+
+Finally, we use that JSON data to plot a time series graph to see how the
+performance changed over time.
+
+### How do we trigger the benchmark run and why GitHub Actions was a bad idea?
+
+To trigger the benchmarkruns we tried using the self hosted runner for GitHub
+Actions but that comes with a few problems of it's own. Anyone who can create a
+fork of the repo can modify the workflow file and run any arbitrary code on our
+machine. [^3]
+
+[^3]: See [github.com/actions/runner/issues/494](https://github.com/actions/runner/issues/494).
+
+> We recommend that you only use self-hosted runners with private repositories.
+> This is because forks of your repository can potentially run dangerous code on
+> your self-hosted runner machine by creating a pull request that executes the
+> code in a workflow. This is not an issue with GitHub-hosted runners because
+> each GitHub-hosted runner is always a clean isolated virtual machine, and it
+> is destroyed at the end of the job execution. Untrusted workflows running on
+> your self-hosted runner pose significant security risks for your machine and
+> network environment, especially if your machine persists its environment
+> between jobs. Some of the risks include:
+>
+> - Malicious programs running on the machine.
+> - Escaping the machine's runner sandbox.
+> - Exposing access to the machine's network environment.
+> - Persisting unwanted or dangerous data on the machine. [^4]
+
+[^4]: [Self-hosted runner security with public repositories](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories).
+
+The day after we were discussing this GitHub announced
+[this](https://github.blog/2021-04-22-github-actions-update-helping-maintainers-combat-bad-actors/)
+which meant that first-time contributors would now require a manual approval
+from maintainers before any Actions workflow can run which solves the problem
+but only partly. This solution assumes that someone has already committed to a
+repository they won’t go rogue which is a fair assumption to make since most
+spammers don't have any contributions to the project they are spamming to but
+still an assumption that can’t be guaranteed.
+
+So in the end we went with a webhooks based solution with a workflow like this:
+
+- A PR gets merged to the main branch or a bench label is added

```

Comment:
What is a bench label?

Reply:
To add benchmarks to PRs I plan to add a bench label so that when the label is added to any PR it is benchmarked and the results are posted as a comment

Reply:
Maybe write it as:

```
or a bench label is added to the PR
```

Reply:
Done :+1: